documentation/projects/proposals/infrastructure_improvements/20240105_implementation_plan_move_remaining_services_to_openverse_org.md

```diff
@@ -0,0 +1,738 @@
+subject to zero-downtime deployment requirements like the API.
+
+This plan is split into two major parts:
+
+1. [Move Airflow to `airflow.openverse.org`](#1-move-airflow-to-airflowopenverseorg)
+2. [Move the Django API to `api.openverse.org` (with staging at `api-staging.openverse.org`)](#2-move-the-django-api-to-apiopenverseorg)
+
+There are also two supplemental parts, a [preliminary step](#0-preliminary-work)
+and a
+[finalisation step](#3-update-references-announce-to-make-blog-and-finalise).
+
+## Expected Outcomes
+
+<!-- List any succinct expected products from this implementation plan. -->
+
+- `openverse.engineering` should move to the Cloudflare free tier
+- All services should live on `openverse.org`
+- Production API domains on `openverse.engineering` should redirect to
+  `openverse.org`
+
+## Step-by-step plan
+
+<!--
+List the ordered steps of the plan in the form of imperative-tone issue titles.
+
+The goal of this section is to give a high-level view of the order of implementation any relationships like
+blockages or other dependencies that exist between steps of the plan. Link each step to the step description
+in the following section.
+
+If special deployments are required between steps, explicitly note them here. Additionally, highlight key
+milestones like when a feature flag could be made available in a particular environment.
+-->
+
+### 0. Preliminary work
+
+1. Move the `access` module that configures Cloudflare Access to a new
+   `cloudflare` root module, maintaining the existing configuration.
+   [The rationale for the new root module is explained in this section](#moving-cloudflare-access-and-the-new-cloudflare-root-module).
+1. Extract load balancer listener rules out of `generic/service` and
+   `generic/ec2-service` in favour of a more generic form of the existing
+   `service-alias` called `service-domain` (effects all existing services in the
+   `next` root modules that have external domains). Update all existing services
+   to use this new `service-domain` module for all domains associated with them.
+1. Similarly, extract security group configuration out of `generic/service` and
+   `generic/ec2-service` into a new module, `generic/security-group`. The new
+   module will configure a default security group with strict access based on
+   our load balancer and SSH-bastion entrypoints for HTTP(S) and SSH
+   respectively. As with the previous point, this deduplicates near identical
+   configuration in `generic/service` and `generic/ec2-service`.
+1. Remove `register_with_target_group` option from `ec2-service` to simplify the
+   module. Registration must be handled using Ansible playbooks for these
+   services, so the updated module should never register the instance to the
+   target group.
+
+### 1. Move Airflow to `airflow.openverse.org`
+
+1. Add `airflow.openverse.org` to Cloudflare access.
+1. Create a new `airflow` module in the `next` root module. Use `ec2-service` as
```

Comment:
With this plan, it seems like we'd have two instances of Airflow running (and accessing the metadata database) at the same time. With the modern version of Airflow we're running, I don't _think_ this should be an issue. The webserver is just a webserver and the scheduler can now [by default have multiple instances running](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/scheduler.html). Just wanted to clarify that this was intended at this step, because at present we remove the current EC2 instance before deploying a new one and only have 1 instance running at any given time.

Reply:
Oh, I should add a step to just spin down the existing one to simplify things. In my mind I definitely meant to just spin down the existing one in the meantime to keep things simple.

Reply:
Sounds good!